### PR TITLE
Clarification updates to IOMMU v1.0.0

### PR DESCRIPTION
**Pull request by ved-rivos**
This PR is opened to accumulate clarification updates to IOMMU v1.0 specification.

### CODE REVIEW

**Comment by ved-rivos** [comment]
@pperesse

**Comment by lyctw** [comment]
hi, there are duplicate entries of `PPN` and `PRI` in the Glossary section

**Comment by ved-rivos** [comment]
Thanks. Will add that to the list.

**Comment by baimengwei** [comment]
If it is convenient for you, can you update the meaning of this sentence below, this sentence is not very easy to understand.

```
The only PTE update supported by the IOMMU without first clearing the V bit in the
original PTE and executing a appropriate `IOTINVAL` command is to do a page size
promotion or demotion.
```

I will think that:
If I want to change the attributes of a PTE, a correct way is as below:
1. modify the attribution position of the PTE
2. set the V position of the PTE to 0
3. execute the IOTINVAL instruction
4. set the V position of the PTE to 1
5. execute the INVAL instruction

However, my usual steps is as follows. This steps is determined by the implementation, and there is no guarantee that these steps will be effective for IOMMUs implemented by other designers.

1. modify the attribution position of the PTE
2. execute the INVAL instruction

Is this understanding correct?

thank you very much.

**Comment by ved-rivos** [comment]
There does not seem a need to make any updates to that sentence. Doing a page size promotion or demotion may be done without marking the PTE invalid. Doing any other changes needs a break-before-make sequence - make PTE invalid, invalidate IOATC, update PTE, and mark valid.

**Comment by baimengwei** [comment]
> There does not seem a need to make any updates to that sentence. Doing a page size promotion or demotion may be done without marking the PTE invalid. Doing any other changes needs a break-before-make sequence - make PTE invalid, invalidate IOATC, update PTE, and mark valid.


Is the software using such an operation below because the hardware may also modify the PTE, causing the software to fail to modify the attributes of the PTE?
```
make PTE invalid, invalidate IOATC, update PTE, and mark valid.
```
The following is a scenario:
1. the software want to write 0 to the position D of the PTE, so the software write 0 to the position D of the PTE.
2. the hardware implicitly modifies the position D of the PTE to 1.
3. software executes IOTINVAL, the position D in the PTE becomes 1 instead of 0, causing the operation to fail, but using the method `make PTE invalid, invalidate IOATC` seems to be ok to avoid failure.

If this is the case, why not check bit 24 of the capabilities register, if this bit is 0, it means that the hardware will not be modified, and then the software only needs to `directly modify the PTE and execute the IOTINVAL instruction`?

**Comment by baimengwei** [comment]
There are some jumps about chapter 2.1.3.6 in the PDF, but the pdf itself does not mark these chapters.

Could you please mark them for easy browsing?

image

image

**Comment by ved-rivos** [comment]
That is covered by the first paragraph of that section - changing A/D bits is allowed without needing PTEs to be marked invalid. The text referenced is in the section about promoting or demoting page size.

**Comment by baimengwei** [comment]
> That is covered by the first paragraph of that section - changing A/D bits is allowed without needing PTEs to be marked invalid. The text referenced is in the section about promoting or demoting page size.

I really can't find the reason for this process to `first clear the V position in PTE`, then ` invalidate the IOATC`, then` modify PTE `, then `set the V position to 1`, and finally execute `invalidate the IOATC ` again. My prior knowledge is that one ` invalidate the IOATC` instruction seems to be enough even for promoting or demoting page size, just as mmu operation in core.😳

**Comment by ved-rivos** [comment]
"and finally execute invalidate the IOATC  again."
where do you see that mentioned?

**Comment by baimengwei** [comment]
> "and finally execute invalidate the IOATC again." where do you see that mentioned?

It was not mentioned, and my understanding of the previous use about mmu is incorrect. I am sorry to bother you. These are just regular operations.



> There does not seem a need to make any updates to that sentence. Doing a page size promotion or demotion may be done without marking the PTE invalid. Doing any other changes needs a break-before-make sequence - make PTE invalid, invalidate IOATC, update PTE, and mark valid.

**Comment by baimengwei** [comment]
is [this ](https://github.com/riscv-non-isa/riscv-iommu/blob/main/iommu_registers.adoc#device-directory-table-pointer-ddtp)a typo? If yes, please modify it. Thanks.

original:
```
When an IOMMU is transitioned to Bare of Off state, the IOMMU may retain information ...
```
expect:
```
When an IOMMU is transitioned to Bare or Off state, the IOMMU may retain information ...
```

**Comment by ved-rivos** [comment]
thanks. Its included in the PR now.

**Comment by baimengwei** [comment]
It seems that [there ](https://github.com/riscv-non-isa/riscv-iommu/blob/main/iommu_registers.adoc?plain=1#L1120)is an extra `S` in `IPSR_FIELDS`, which leads to a jump failure in PDF.

```
...
[[IPSR_FIELD]]
...

If the conditions to set that bit are still present (See <<IPSR_FIELDS>>) or if
...
```
image

**Comment by ved-rivos** [comment]
Thanks. FIxed IPSR_FIELDS->IPSR_FIELD.

**Comment by ved-rivos** [comment]
> Could you please mark them with number for easy browsing?

This is included in the PR.

**Comment by baimengwei** [comment]
Configure the address translation of MSI to pass-through mode, it seems that there is no PTE of MSI according to the `DC` format, but [there ](https://github.com/riscv-non-isa/riscv-iommu/blob/main/iommu_registers.adoc#CAP) is a `MSI PTE is supported` in the following sentence. It seems that `MSI FLAT` and `MSI PTE is supported` are not related?
```
22 | MSI_FLAT | RO | MSI address translation using Pass-through mode MSI PTE is supported.
```

**Comment by baimengwei** [comment]
It would be better if the first letters are in the same capitalization style.
image

**Review by pperesse — APPROVED** [review]
LGTM - thanks

**Review by gajinderpanesar — APPROVED** [review]
Looks fine to me

**Review by gajinderpanesar — APPROVED** [review]
Looks fine

**Review by gajinderpanesar — APPROVED** [review]
Looks fine

File: iommu_in_memory_queues.adoc [hunk 1]
```diff
@@ -270,6 +270,11 @@ match the `GSCID` argument, regardless of the address argument.
 Simpler implementations may ignore the operand of `IOTINVAL.VMA` and/or
 `IOTINVAL.GVMA` and always perform a global invalidation of all
 address-translation entries.
+
+Some implementations may choose to cache an identity-mapped translation for the
+stage of address translation that is in `Bare` mode. The 'IOTINVAL' commands do
+not apply to such identity-mapped translations, as these identity mappings are
```
Comment (ved-rivos) [inline iommu_in_memory_queues.adoc:279]: -> do not need to apply.

File: iommu_in_memory_queues.adoc [hunk 2]
```diff
@@ -180,7 +180,10 @@ operand is valid. Setting `PSCV` to 1 is allowed only for `IOTINVAL.VMA`. The
 the translations associated with the host (i.e. those where the second-stage
 is Bare) are operated on. When `GV` is 0, the `GSCID` operand is ignored.
 When `AV` is 0, the `ADDR` operand is ignored. When `PSCV` operand is 0, the
-`PSCID` operand is ignored.
+`PSCID` operand is ignored. When the `AV` operand is set to 1, if the `ADDR`
+operand specifies an invalid address, the implementation may either complete the
+command with no effect or may complete the command using an alternate, yet
+`UNSPECIFIED`, legal value for the `ADDR` operand.
```
Comment (ved-rivos) [inline iommu_in_memory_queues.adoc:186]: When the `AV` operand is set to 1, if the `ADDR` operand specifies an invalid address, the command may or may not perform any invalidations.
note:
An implementation may either complete the command with no effect or may complete the command using an alternate, yet `UNSPECIFIED`, legal value for the `ADDR` operand. Note that  entries in general may be invalidated from the address translation cache at any time.

File: iommu_registers.adoc [hunk 3]
```diff
@@ -234,6 +234,9 @@ must be supported.
 IOMMU implementations must support the Svnapot standard extension for NAPOT
 Translation Contiguity.
 
+The `PAS` field specifies the width, in bits, of the physical address supported
+by the IOMMU.
+
```
Comment (ved-rivos) [inline iommu_registers.adoc:241]: The physical address space addressable by the IOMMU is 0 to 2^(capabilities.PAS)-1.

File: iommu_registers.adoc [hunk 4]
```diff
@@ -68,8 +68,10 @@ the register returns 0 and writes to that offset are ignored.
                                        CSR >>             | if `capabilities.ATS==0`
 |84    |`ipsr`          |4   |<<IPSR, Interrupt pending
                                          status register>>| No
-|88    |`iocntovf`      |4   |<<OVF, HPM counter overflows>> | if `capabilities.HPM==0`
-|92    |`iocntinh`      |4   |<<INH, HPM counter inhibits>> | if `capabilities.HPM==0`
+|88    |`iocountovf`    |4   |<<OVF, HPM counter
```
Comment (jones-drew) [inline iommu_registers.adoc:71]: There is still another reference to iocntovf in the iocountovf section.

Comment (ved-rivos) [reply iommu_registers.adoc:71]: Fixed. Thanks!

**Comment by jones-drew** [comment]
Section 3.1 lists the opcode 3 as IOTDIR, but then later refers to that opcode as IODIR in section 3.1.3.

**Comment by ved-rivos** [comment]
@jones-drew that correction is already in this PR

**Comment by jones-drew** [comment]
> @jones-drew that correction is already in this PR

Thanks and sorry I didn't check the full diff first!

**Pull request merged**